driver-sync/src/main/com/mongodb/client/internal/AggregateIterableImpl.java

```diff
@@ -169,21 +169,34 @@ public void toCollection() {
             if (batchSize != null) {
                 findOptions.batchSize(batchSize);
             }
-            return operations.find(new MongoNamespace(namespace.getDatabaseName(), outCollection.asString().getValue()), new BsonDocument(),
-                    resultClass, findOptions);
+            return operations.find(outNamespace, new BsonDocument(), resultClass, findOptions);
         } else {
             return operations.aggregate(pipeline, resultClass, maxTimeMS, maxAwaitTimeMS, getBatchSize(), collation,
                     hint, comment, allowDiskUse, useCursor, aggregationLevel);
         }
     }
 
     @Nullable
-    private BsonValue getOutCollection() {
+    private MongoNamespace getOutNamespace() {
         if (pipeline.size() == 0) {
             return null;
         }
 
         Bson lastStage = notNull("last stage", pipeline.get(pipeline.size() - 1));
-        return lastStage.toBsonDocument(documentClass, codecRegistry).get("$out");
+        BsonDocument lastStageDocument = lastStage.toBsonDocument(documentClass, codecRegistry);
```

Comment:
Hate this repeated code, but we don't have a good way to share stuff like this between sync and async.